### PR TITLE
Add granular permissions support for Gateway operator users

### PR DESCRIPTION
**Pull request by zeeshanakram3**
Addresses #218 

Specification: #169

# How to use the permissions mutations

- The root gateway user can use `grantPermissions` (to grant any operator permissions) or `revokePermissions` (to revoke any operator permissions) for an existing gateway user.
- The root gateway operator first needs to be authorized using Orion Auth API's `/anonymous-auth` endpoint, after that they can go to the graphql playground to execute the mutations like this:

### Using `grantPermissions` mutation

image

### Using `revokePermissions` mutation

image

**_Ideally, the Root user can create a few gateway users (as many as required) and execute `grantPermissions` or `revokePermissions` against those `userId`/s. Then share those `userId`s with relevant people privately (as the `userId` will act as operator secret so it's important to keep it secret)_**

### CODE REVIEW

File: src/server-extension/resolvers/AdminResolver/index.ts [hunk 1]
```diff
@@ -24,36 +50,45 @@ import {
...
+
+    // Add only new permissions that the user doesn't have yet
+    user.permissions = Array.from(new Set([...(user.permissions || []), ...args.permissions]))
+
+    await em.save(user)
+    return true
+  }
+
+  @UseMiddleware(OperatorOnly(OperatorPermission.REVOKE_OPERATOR_PERMISSIONS))
+  @Mutation(() => Boolean)
```
Comment (ignazio-bovo) [inline src/server-extension/resolvers/AdminResolver/index.ts:77]: Maybe return the new user permission set?
From a user perspective I guess the raw boolean is a bit cryptic, I guess, but I can be wrong, plus all the `set*` mutation that we already have returns the new value just set eg `{isKilled: bool}` for the `setKillswitch` mutation

File: src/server-extension/resolvers/AdminResolver/index.ts [hunk 2]
```diff
@@ -24,36 +50,45 @@ import {
...
+  async revokePermission(@Args() args: RevokeOperatorPermissionsInput): Promise<boolean> {
```
Comment (ignazio-bovo) [inline src/server-extension/resolvers/AdminResolver/index.ts:78]: Same comment as previous:
Maybe return the new permission set here? WDYT?

**Review by ignazio-bovo — CHANGES_REQUESTED** [review]
This is ok, I have no major issues in approving the PR as is
I left some minor improvements

File: docker-compose.yml [hunk 3]
```diff
@@ -10,8 +10,7 @@ services:
       POSTGRES_DB: squid
       POSTGRES_PASSWORD: squid
     ports:
-      - '127.0.0.1:${DB_PORT}:${DB_PORT}'
-      - '[::1]:${DB_PORT}:${DB_PORT}'
+      - '${DB_PORT}:${DB_PORT}'
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply docker-compose.yml:13]: Yes, this is a mistake. I will revert it. Apparently this port scheme was not working for me locally in Mac.

File: schema/auth.graphql [hunk 4]
```diff
@@ -1,10 +1,27 @@
...
+
 type User @entity {
   "Unique identifier (32-byte string, securely random)"
   id: ID!
 
   "Whether the user has root (gateway operator) privileges"
   isRoot: Boolean!
 
+  "List of all the gateway operator permissions that this user has"
+  permissions: [OperatorPermission!]
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply schema/auth.graphql:23]: I was thinking in terms of migration ease. i.e. if it's a nullable field then we wont have to prepare a migration to set emply `[]` for all the existing users? WDYT?

File: src/server-extension/resolvers/AdminResolver/index.ts [hunk 5]
```diff
@@ -24,36 +50,45 @@ import {
...
+  @Mutation(() => Boolean)
+  async grantPermissions(@Args() args: GrantOperatorPermissionsInput): Promise<boolean> {
+    const em = await this.em()
+    const user = await em.findOne(User, { where: { id: args.userId } })
+    if (!user) {
+      throw new Error('User not found')
+    }
+
+    // Add only new permissions that the user doesn't have yet
+    user.permissions = Array.from(new Set([...(user.permissions || []), ...args.permissions]))
```
[the remark this answers is not in the document]
Comment (zeeshanakram3) [reply src/server-extension/resolvers/AdminResolver/index.ts:70]: Please see the above comment

File: schema/auth.graphql (hunk 4 above)
[the remark this answers is not in the document]
Comment (ignazio-bovo) [reply schema/auth.graphql:23]: ok then I am ok with this

File: src/server-extension/resolvers/AdminResolver/index.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ignazio-bovo) [reply src/server-extension/resolvers/AdminResolver/index.ts:73]: ok then I am ok with this

**Review by ignazio-bovo — APPROVED** [review]
LGTM

**Pull request merged**